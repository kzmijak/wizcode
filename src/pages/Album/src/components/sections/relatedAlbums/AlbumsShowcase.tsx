import { Stack, Tooltip } from "@mui/joy";
import { Picture } from "components/Picture";
import { AlbumModel } from "models/Album";
import { FC } from "react";
import { Link } from "react-router-dom";
import { PATH_APP } from "routes/src/paths";

type AlbumsShowcaseProps = { albums: AlbumModel[] };

export const AlbumsShowcase: FC<AlbumsShowcaseProps> = ({ albums }) => {
  return (
    <Stack spacing={2} component="section" direction="row" flexWrap="nowrap">
      {albums.map((album) => (
        <Tooltip key={album.id} title={album.title}>
          <Stack
            component={Link}
            to={PATH_APP.album + album.id}
            sx={(theme) => ({
              backgroundColor: theme.palette.background.level1,
            })}
            direction="row"
          >
            <Picture
              src={album.imageUrl}
              alt={`${album.title} - related album cover`}
              width={55}
              height={55}
            />
          </Stack>
        </Tooltip>
      ))}
    </Stack>
  );
};
