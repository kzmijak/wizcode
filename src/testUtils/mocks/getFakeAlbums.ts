import { AlbumModel } from "models/Album";

export const getFakeAlbums = (count: number) =>
  [...Array(count).keys()].map<AlbumModel>((key) => ({
    artistName: "artist" + key,
    category: "category" + key,
    id: key.toString(),
    imageUrl: key.toString(),
    priceInt: key * 100,
    releaseDate: new Date(key),
    songsCount: key,
    title: "title" + key,
  }));
