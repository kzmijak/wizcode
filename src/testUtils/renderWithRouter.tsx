import { render } from "@testing-library/react";
import { RouterProvider, createMemoryRouter } from "react-router";

export const renderWithRouter = (ui: JSX.Element, ...location: string[]) => {
  const router = createMemoryRouter([{ path: "*", element: ui }], {
    initialEntries: ["", ...location],
  });

  return render(<RouterProvider router={router} />);
};
