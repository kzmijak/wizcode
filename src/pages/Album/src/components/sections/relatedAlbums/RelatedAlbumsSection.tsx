import { Button, Divider, Stack, Typography } from "@mui/joy";
import { AlbumModel } from "models/Album";
import { FC } from "react";
import { AlbumsShowcase } from "./AlbumsShowcase";
import { Link } from "react-router-dom";
import { PATH_APP } from "routes/src/paths";
import { albumsByCategoryState } from "modules/AlbumsState";
import { useRecoilValue } from "recoil";

type RelatedAlbumsSectionProps = Pick<AlbumModel, "id" | "category">;

export const RelatedAlbumsSection: FC<RelatedAlbumsSectionProps> = (album) => {
  const sameCategoryAlbums = useRecoilValue(
    albumsByCategoryState(album.category)
  );
  const relatedAlbums = sameCategoryAlbums.filter(
    (sameCategoryAlbum) => sameCategoryAlbum.id !== album.id
  );
  relatedAlbums.splice(14);

  if (relatedAlbums.length === 0) return null;
  return (
    <>
      <Typography level="h4">Other albums you may like</Typography>
      <Divider />
      <Stack direction="row" justifyContent="space-between">
        <AlbumsShowcase albums={relatedAlbums} />
        <Button component={Link} to={PATH_APP.root}>
          Show all
        </Button>
      </Stack>
    </>
  );
};
