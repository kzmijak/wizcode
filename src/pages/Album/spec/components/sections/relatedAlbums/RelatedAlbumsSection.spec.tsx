import { describe, it, expect, vi } from "vitest";
import { screen } from "@testing-library/react";
import { RelatedAlbumsSection } from "../../../../src/components/sections/relatedAlbums/RelatedAlbumsSection";
import * as recoil from "recoil";
import { renderWithRouter } from "testUtils/renderWithRouter";
import { getFakeAlbums } from "testUtils/mocks/getFakeAlbums";

vi.mock("recoil", async () => {
  const lib = (await vi.importActual("recoil")) as object;
  return {
    ...lib,
    useRecoilValue: vi.fn(),
  };
});

const albums = getFakeAlbums(20);

describe("RelatedAlbumsSection", () => {
  const useRecoilValueMocked = vi.mocked(recoil.useRecoilValue);

  afterEach(() => {
    useRecoilValueMocked.mockReset();
  });

  const Component = () => (
    <RelatedAlbumsSection category="pop" id="thisAlbum" />
  );

  it("should match snapshot", () => {
    useRecoilValueMocked.mockReturnValue([albums[0]]);
    expect(renderWithRouter(<Component />)).toMatchSnapshot();
  });

  it("should render header", () => {
    useRecoilValueMocked.mockReturnValue([albums[0]]);
    renderWithRouter(<Component />);

    expect(screen.queryByText("Other albums you may like")).toBeInTheDocument();
  });

  it("should render nothing, when na related albums available", () => {
    useRecoilValueMocked.mockReturnValue([]);
    renderWithRouter(<Component />);

    expect(
      screen.queryByText("Other albums you may like")
    ).not.toBeInTheDocument();
  });

  it("should only display up to 14 albums at a time", () => {
    useRecoilValueMocked.mockReturnValue(albums);
    renderWithRouter(<Component />);

    expect(screen.queryAllByRole("img").length).toBe(14);
  });

  it("should contain show all button", () => {
    useRecoilValueMocked.mockReturnValue(albums);
    renderWithRouter(<Component />);

    const showAllButton = screen.queryByText("Show all");
    expect(showAllButton).toBeInTheDocument();
    expect(showAllButton).toHaveAttribute("href", "/");
  });
});
