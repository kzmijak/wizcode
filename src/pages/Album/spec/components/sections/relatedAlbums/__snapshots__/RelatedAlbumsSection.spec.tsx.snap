// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`RelatedAlbumsSection > should match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <h4
        class="MuiTypography-root MuiTypography-h4 css-nk02qn-JoyTypography-root"
      >
        Other albums you may like
      </h4>
      <hr
        class="MuiDivider-root MuiDivider-horizontal css-10w8je6-JoyDivider-root"
      />
      <div
        class="MuiStack-root css-ipxlt3-JoyStack-root"
      >
        <section
          class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-row MuiGrid-spacing-xs-2 css-18ev1m-JoyGrid-root"
        >
          <a
            aria-label="title0"
            class="MuiGrid-root MuiGrid-direction-xs-row css-bqntaw-JoyGrid-root"
            href="/0"
          >
            <div
              class="MuiStack-root css-f9pi93-JoyStack-root"
            >
              <span
                class="MuiCircularProgress-root MuiCircularProgress-colorPrimary MuiCircularProgress-variantSoft MuiCircularProgress-sizeMd css-sdm6f4-JoyCircularProgress-root"
                role="progressbar"
                style="--CircularProgress-percent: 25;"
              >
                <svg
                  class="MuiCircularProgress-svg css-qw6zo9-JoyCircularProgress-svg"
                >
                  <circle
                    class="MuiCircularProgress-track css-vtf6ih-JoyCircularProgress-track"
                  />
                  <circle
                    class="MuiCircularProgress-progress css-glg1l6-JoyCircularProgress-progress"
                  />
                </svg>
              </span>
              <img
                alt="title0 - related album cover"
                src="0"
                style="opacity: 0; height: 55px; width: 55px;"
              />
            </div>
          </a>
        </section>
        <a
          class="MuiButton-root MuiButton-variantSolid MuiButton-colorPrimary MuiButton-sizeMd css-grtnkv-JoyButton-root"
          href="/"
        >
          Show all
        </a>
      </div>
    </div>
  </body>,
  "container": <div>
    <h4
      class="MuiTypography-root MuiTypography-h4 css-nk02qn-JoyTypography-root"
    >
      Other albums you may like
    </h4>
    <hr
      class="MuiDivider-root MuiDivider-horizontal css-10w8je6-JoyDivider-root"
    />
    <div
      class="MuiStack-root css-ipxlt3-JoyStack-root"
    >
      <section
        class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-row MuiGrid-spacing-xs-2 css-18ev1m-JoyGrid-root"
      >
        <a
          aria-label="title0"
          class="MuiGrid-root MuiGrid-direction-xs-row css-bqntaw-JoyGrid-root"
          href="/0"
        >
          <div
            class="MuiStack-root css-f9pi93-JoyStack-root"
          >
            <span
              class="MuiCircularProgress-root MuiCircularProgress-colorPrimary MuiCircularProgress-variantSoft MuiCircularProgress-sizeMd css-sdm6f4-JoyCircularProgress-root"
              role="progressbar"
              style="--CircularProgress-percent: 25;"
            >
              <svg
                class="MuiCircularProgress-svg css-qw6zo9-JoyCircularProgress-svg"
              >
                <circle
                  class="MuiCircularProgress-track css-vtf6ih-JoyCircularProgress-track"
                />
                <circle
                  class="MuiCircularProgress-progress css-glg1l6-JoyCircularProgress-progress"
                />
              </svg>
            </span>
            <img
              alt="title0 - related album cover"
              src="0"
              style="opacity: 0; height: 55px; width: 55px;"
            />
          </div>
        </a>
      </section>
      <a
        class="MuiButton-root MuiButton-variantSolid MuiButton-colorPrimary MuiButton-sizeMd css-grtnkv-JoyButton-root"
        href="/"
      >
        Show all
      </a>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
